### PR TITLE
Aliu/fd tunneling pr

### PR DESCRIPTION
**Pull request by aliuTT**
This is working: host sending FD cmds/data --> remote issue core --> eth SRC router --> eth DST router. 

### Features overview:
- Added some cluster apis to enable and disable internal routing (slow and fast dispatch depending on env). Enable should be called after all devices are initialized, disable should be called right before all devices are closed. We should make this into some user facing cluster handle.
- Eth kernels, integrated with Almeet's remote issue queue changes. Split ethernet dataflow apis into user facing and `internal_` namespaced apis.
- Refactored dispatch kernels to use an L1 struct for db cb configs.

### CODE REVIEW

File: tt_metal/hw/inc/ethernet/dataflow_api.h [hunk 1]
```diff
@@ -38,20 +40,14 @@ volatile uint32_t *RtosTable =
 
 void (*rtos_context_switch_ptr)();
 
+namespace internal_ {
```
Comment (aliuTT) [inline tt_metal/hw/inc/ethernet/dataflow_api.h:43]: @tt-aho namespacing some functions that should only be called from within ethernet dataflow implementation

File: tt_metal/impl/dispatch/kernels/command_queue_common.hpp [hunk 2]
```diff
@@ -5,67 +5,29 @@
 #include "dataflow_api.h"
 #include "tt_metal/impl/dispatch/device_command.hpp"
 
-static constexpr uint32_t l1_db_cb_addr_offset = 7 * 16;
```
Comment (aliuTT) [inline tt_metal/impl/dispatch/kernels/command_queue_common.hpp:8]: @pgkeller @DrJessop Refactored the following into an L1 struct

File: tt_metal/detail/tt_metal.hpp [hunk 3]
```diff
@@ -442,95 +446,134 @@ namespace tt::tt_metal{
...
+
+                        CoreCoord issue_q_logical_core(issue_q_interface_location.x, issue_q_interface_location.y);
+                        CoreCoord completion_q_logical_core(completion_q_interface_location.x, completion_q_interface_location.y);
+                        CoreCoord issue_q_physical_core = get_physical_core_coordinate(issue_q_interface_location, CoreType::WORKER);
+                        CoreCoord completion_q_physical_core = get_physical_core_coordinate(completion_q_interface_location, CoreType::WORKER);
+
+                        if (device_id != device->id()) {
+                            // This means the issue queue and completion queue interfaces that service a remote device are being set up
+                            // the issue queue interface needs to send fast dispatch packets to the "src" ethernet core
+                            CoreCoord logical_eth_router_src =
```
Comment (aliuTT) [inline tt_metal/detail/tt_metal.hpp:477]: @abhullar-tt getting the corresponding eth router cores here now

File: tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp [hunk 4]
```diff
@@ -33,9 +33,11 @@ class DeviceFixture : public ::testing::Test {
             auto* device = tt::tt_metal::CreateDevice(id);
             devices_.push_back(device);
         }
```
Comment (aliuTT) [inline tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: @abhullar-tt calling these cluster enable/disable of the internal routing

File: tt_metal/detail/tt_metal.hpp [hunk 5]
```diff
@@ -553,7 +553,14 @@ namespace tt::tt_metal{
...
                             // the issue queue interface needs to send fast dispatch packets to the "src" ethernet core
-                            consumer_physical_core = device->ethernet_core_from_logical_core(*device->get_active_ethernet_cores().begin());
+                            CoreCoord logical_eth_router_src =
+                                tt::Cluster::instance().get_and_configure_corresponding_eth_core_for_device(
+                                    device->id(), issue_q_physical_core, EthRouterMode::FD_SRC, device_id);
+                            consumer_physical_core = device->ethernet_core_from_logical_core(logical_eth_router_src);
+                            // TODO: remote processor core should use logical_eth_router_dst as producer
+                            CoreCoord logical_eth_router_dst =
+                                tt::Cluster::instance().get_and_configure_corresponding_eth_core_for_device(
+                                    device_id, issue_q_physical_core, EthRouterMode::FD_DST, device->id());
```
Comment (abhullar-tt) [inline tt_metal/detail/tt_metal.hpp:563]: you can remove this from here, if you look above this case is for programming the MMIO device with dispatch kernels. I can grab the eth dst router when setting up the program for the remote device

File: tt_metal/hw/firmware/src/erisc.cc [hunk 6]
```diff
@@ -28,13 +28,50 @@ uint32_t wIndex __attribute__((used));
...
+uint8_t noc_index = 0;  // TODO: remove hardcoding
 uint8_t my_x[NUM_NOCS] __attribute__((used));
 uint8_t my_y[NUM_NOCS] __attribute__((used));
 
 uint32_t noc_reads_num_issued[NUM_NOCS] __attribute__((used));
 uint32_t noc_nonposted_writes_num_issued[NUM_NOCS] __attribute__((used));
 uint32_t noc_nonposted_writes_acked[NUM_NOCS] __attribute__((used));
 
+FORCE_INLINE
+void multicore_eth_cb_wait_front(db_cb_config_t *eth_db_cb_config, int32_t num_pages) {
```
Comment (abhullar-tt) [inline tt_metal/hw/firmware/src/erisc.cc:40]: I might modify this to create multicore cb apis that work for reader/writer being different core types. since we need this sync mechanism for multiple dispatch cores

File: tt_metal/hw/firmware/src/erisc.cc [hunk 7]
```diff
@@ -54,14 +91,65 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
...
+                const uint32_t num_pages = command_ptr[2];
+                uint32_t num_writes_completed = 0;
+                while (num_writes_completed != num_pages) {
+                    uint32_t num_to_write = min(num_pages, producer_consumer_transfer_num_pages);
+                    multicore_eth_cb_wait_front(eth_db_cb_config, num_to_write);
+                    // contains device command, maybe just send pages, and send cmd once at the start
+                    internal_::send_fd_packets();
+                    multicore_eth_cb_pop_front(
+                        eth_db_cb_config, remote_db_cb_config, dispatch_core_noc_encoding, num_to_write);
+                    num_writes_completed += num_to_write;
```
Comment (abhullar-tt) [inline tt_metal/hw/firmware/src/erisc.cc:141]: so the cbs handle how many pages of data are being sent / received but the first packet being sent to the dst router  contains the device command as well?

File: tt_metal/detail/tt_metal.hpp (hunk 5 above)
Comment (aliuTT) [reply tt_metal/detail/tt_metal.hpp:563]: yea I was thinking about this.. what happens if we have four ethernet links like we do in Galaxy, and between two chips we have two cores as SRC and two as DST. Wouldn't it be more correct to somehow set up the DST by just grabbing the connected ethernet core for the selected eth SRC?

File: tt_metal/impl/device/device.hpp [hunk 8]
```diff
@@ -117,6 +117,12 @@ class Device {
         return tt::Cluster::instance().get_connected_ethernet_core(std::make_tuple(this->id_, eth_core));
     }
 
+    CoreCoord get_and_configure_corresponding_eth_core(
+        CoreCoord physical_dispatch_core, EthRouterMode mode, chip_id_t connected_chip_id) const {
+        return tt::Cluster::instance().get_and_configure_corresponding_eth_core_for_device(
```
Comment (abhullar-tt) [inline tt_metal/impl/device/device.hpp:122]: most of our apis accept logical core coordinates. also should `CoreCoord physical_dispatch_core` be `tt_cxy_pair logical_dispatch_location` and then get the connected chip based off the chip in `logical_dispatch_location`?

File: tt_metal/hw/firmware/src/erisc.cc (hunk 6 above)
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:40]: I think we need `_eth_` variants of every noc stall, because we need to call a context switch in the while loops if we want DPRINT to flush for remote chips during debug.

File: tt_metal/impl/device/device.hpp (hunk 8 above)
Comment (abhullar-tt) [reply tt_metal/impl/device/device.hpp:122]: also should we break these up into two sep apis, one to get the eth core and another to configure it?

File: tt_metal/hw/firmware/src/erisc.cc (hunk 6 above)
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:40]: I just haven't added the context switch here yet

File: tt_metal/detail/tt_metal.hpp (hunk 5 above)
Comment (abhullar-tt) [reply tt_metal/detail/tt_metal.hpp:563]: when we configure the remote dispatch program we can find which chip the remote issue queue reader is on and then get the eth router associated with that core which will give us src router. From the src we can get the corresponding dst? and ensure this matches to the eth router that feeds the remote processor

File: tt_metal/hw/firmware/src/erisc.cc (hunk 7 above)
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:141]: every packet is sending the full eth FD reserved L1, so it contains cmd + data + junk L1. This is because I measured the perf of sending eth packets, and my hunch is that it's more performant to send one transaction request for a larger L1 than add risc logic to offset addresses and issue smaller L1 transfers. This is something we can tweak later though

File: tt_metal/impl/device/device.hpp (hunk 8 above)
Comment (aliuTT) [reply tt_metal/impl/device/device.hpp:122]: I can change the core coord to be logical and group the connected chip to `tt_cxy_pair`. I think breaking this to two apis makes sense, since I guess we might want to call the getter in other places.

File: tt_metal/hw/inc/ethernet/dataflow_api.h [hunk 9]
```diff
@@ -65,6 +61,63 @@ void check_and_context_switch() {
...
+        1);
+}
+
+}  // namespace internal_
+
+void run_routing() {
+    // TODO: maybe split into two FWs? or this may be better to sometimes allow each eth core to do both send and
+    // receive of fd packets
+    if (routing_info->routing_mode == EthRouterMode::FD_SRC) {
+        // TODO: port changes from erisc to here
```
Comment (abhullar-tt) [inline tt_metal/hw/inc/ethernet/dataflow_api.h:109]: is this TODO part of this PR?

File: tt_metal/llrt/tt_cluster.hpp [hunk 10]
```diff
@@ -119,6 +119,20 @@ class Cluster {
...
+    // Configures routing mapping of ethernet cores
+    void initialize_routing_info_for_ethernet_cores();
+
+    // Dispatch core is managed by device, so this is an api for device to write the routing info to each eth core used
+    // in FD tunneling Returns logical eth core that communicates with specified dispatch core
+    CoreCoord get_and_configure_corresponding_eth_core_for_device(
+        chip_id_t chip_id, CoreCoord physical_dispatch_core, EthRouterMode mode, chip_id_t connected_chip_id) const;
+
+    // Writes eth routing config to device, assumes that device is initialized
+    void set_internal_routing_info_for_ethernet_cores(bool enable_internal_routing) const;
```
Comment (abhullar-tt) [inline tt_metal/llrt/tt_cluster.hpp:134]: can you add comment here explaining what internal routing is

File: tt_metal/hw/inc/ethernet/dataflow_api.h (hunk 9 above)
Comment (aliuTT) [reply tt_metal/hw/inc/ethernet/dataflow_api.h:109]: No, it's more refactor work so things look nicer. I may clean it up in the next phase of integrating with your remote processor changes. I just avoided refactoring in case this dramatically changes

File: tt_metal/llrt/tt_cluster.hpp (hunk 10 above)
Comment (aliuTT) [reply tt_metal/llrt/tt_cluster.hpp:134]: Will do

File: tt_metal/hw/inc/dataflow_api.h [hunk 11]
```diff
@@ -1151,6 +1151,11 @@ uint32_t get_semaphore(uint32_t semaphore_id) {
     return SEMAPHORE_BASE + semaphore_id * sizeof(uint32_t);
 }
 
+FORCE_INLINE
+uint32_t eth_get_semaphore(uint32_t semaphore_id) {
+    return eth_l1_mem::address_map::SEMAPHORE_BASE + semaphore_id * sizeof(uint32_t);
```
Comment (DrJessop) [inline tt_metal/hw/inc/dataflow_api.h:1156]: Almeet pushed a bug fix related to this. See her fix https://github.com/tenstorrent-metal/tt-metal/pull/4820

File: tt_metal/hw/inc/dev_msgs.h [hunk 12]
```diff
@@ -111,3 +111,20 @@ static_assert(MEM_MAILBOX_BASE + offsetof(mailboxes_t, slave_sync.ncrisc) == MEM
...
+
+struct routing_info_t {
+    volatile uint32_t routing_enabled;
+    volatile uint32_t routing_mode;
+    volatile uint32_t connected_chip_id;
+    volatile uint32_t dispatch_core_xy;
+    volatile uint32_t fd_buffer_msgs_sent;
+    uint32_t reserved_0_;
+    uint32_t reserved_1_;
+    uint32_t reserved_2_;
```
Comment (DrJessop) [inline tt_metal/hw/inc/dev_msgs.h:129]: why the post-fix underscore?

File: tt_metal/hw/inc/grayskull/eth_l1_address_map.h [hunk 13]
```diff
@@ -16,14 +16,18 @@ struct address_map {
...
+  static constexpr std::int32_t ERISC_APP_ROUTING_INFO_BASE = 0;
   static constexpr std::int32_t ERISC_APP_SYNC_INFO_BASE = 0;
   static constexpr std::int32_t ERISC_L1_ARG_BASE = 0;
 
   static constexpr std::int32_t ERISC_FIRMWARE_SIZE = 16;
   static constexpr std::int32_t ERISC_APP_RESERVED_BASE = 0;
   static constexpr std::int32_t ERISC_APP_RESERVED_SIZE = 16;
   static constexpr std::int32_t ERISC_L1_UNRESERVED_BASE = 0;
+  static constexpr std::uint32_t SEMAPHORE_BASE = 0;
+  static constexpr std::uint32_t CQ_CONSUMER_CB_BASE = 0;
```
Comment (DrJessop) [inline tt_metal/hw/inc/grayskull/eth_l1_address_map.h:30]: why 0s?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp [hunk 14]
```diff
@@ -35,45 +35,36 @@ void noc_async_write_multicast_one_packet_no_path_reserve(
...
+    uint32_t num_to_write,
     uint32_t page_size) {
-    volatile tt_l1_ptr uint32_t* CQ_CONSUMER_CB_ACK_PTR = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_db_cb_ack_addr(db_buf_switch));
-    volatile tt_l1_ptr uint32_t* CQ_CONSUMER_CB_READ_PTR =
-        reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_db_cb_rd_ptr_addr(db_buf_switch));
-
-    *CQ_CONSUMER_CB_ACK_PTR += num_pages;
-    *CQ_CONSUMER_CB_READ_PTR += (page_size * num_pages) >> 4;
+    db_cb_config->ack += num_to_write;
+    db_cb_config->rd_ptr += page_size * num_to_write;
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:56]: why was this changed? page_size is always in units of bytes, not 16B.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp [hunk 15]
```diff
@@ -35,45 +35,36 @@ void noc_async_write_multicast_one_packet_no_path_reserve(
...
 
-    if ((*CQ_CONSUMER_CB_READ_PTR << 4) > fifo_limit) {
-        *CQ_CONSUMER_CB_READ_PTR -= fifo_size >> 4;
+    if ((db_cb_config->rd_ptr << 4) > consumer_fifo_limit) {
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:58]: I think this should be >= now, need to confirm, but I believe the limit is exclusive rather than inclusive now. This could have been a latent bug?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp [hunk 16]
```diff
@@ -152,8 +141,8 @@ FORCE_INLINE void write_buffers(
         const uint32_t dst_page_index = command_ptr[7];
 
         uint32_t num_to_write;
-        uint32_t src_addr = *reinterpret_cast<volatile tt_l1_ptr uint32_t*>(get_db_cb_rd_ptr_addr(db_buf_switch)) << 4;
-        uint32_t l1_consumer_fifo_limit = src_addr + consumer_cb_size - 1;
+        uint32_t src_addr = get_read_ptr(db_cb_config);
+        uint32_t l1_consumer_fifo_limit = src_addr + (db_cb_config->total_size << 4) - 1;
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:145]: I think in general, we make limits be exclusive. Somehow the old inclusive flow stuck around in some cases. We should make it consistent and have it be the exclusive flow.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp [hunk 17]
```diff
@@ -222,46 +211,42 @@ FORCE_INLINE void write_program_page(uint32_t page_addr, volatile tt_l1_ptr uint
...
         multicore_cb_pop_front(
+            db_cb_config,
+            remote_db_cb_config,
             producer_noc_encoding,
-            db_buf_switch,
             l1_consumer_fifo_limit,
-            consumer_cb_size,
             num_to_write,
-            DeviceCommand::PROGRAM_PAGE_SIZE);
+            (DeviceCommand::PROGRAM_PAGE_SIZE >> 4));
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:237]: not a fan of this.

**Review by DrJessop — CHANGES_REQUESTED** [review]
Lots to review, here is my first phase of reviews. Great work, but there are a few things to fix/clean up.

File: tt_metal/hw/firmware/src/erisc.cc [hunk 18]
```diff
@@ -54,14 +91,65 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
...
+    while (routing_info->routing_enabled) {
+        // FD: assume that no more host -> remote writes are pending
+        if (erisc_info->launch_user_kernel == 1) {
             kernel_init();
-            break;
+        } else if (routing_info->routing_mode == EthRouterMode::FD_SRC) {
+            eth_db_acquire(eth_db_semaphore_addr, eth_router_noc_encoding);
+            if (!routing_info->routing_enabled) {
+                break;
+            }
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:114]: Host sends disable when cluster is closed.

File: tt_metal/hw/firmware/src/erisc.cc [hunk 19]
```diff
@@ -54,14 +91,65 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
...
+    }
+    uint64_t dispatch_core_noc_encoding = get_dispatch_core_noc_encoding();
+    uint64_t eth_router_noc_encoding = uint64_t(NOC_XY_ENCODING(my_x[0], my_y[0])) << 32;
+
+    volatile tt_l1_ptr uint32_t *eth_db_semaphore_addr =
+        reinterpret_cast<volatile tt_l1_ptr uint32_t *>(eth_get_semaphore(0));
+
+    static constexpr uint32_t command_start_addr = eth_l1_mem::address_map::ERISC_APP_RESERVED_BASE;
+
+    bool db_buf_switch = false;
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:105]: For all dispatch kernels? I have a feeling we will eventually want to double buffer on ethernet cores too.

File: tt_metal/hw/inc/grayskull/eth_l1_address_map.h (hunk 13 above)
Comment (aliuTT) [reply tt_metal/hw/inc/grayskull/eth_l1_address_map.h:30]: To fix GS builds, since eth FW is built by default for all arches. I guess deleting this file for GS can be another cleanup item?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp (hunk 14 above)
Comment (aliuTT) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:56]: Hmm I was just refactoring here, but the page size was 16B sized from before (this should be on main):
```
- reinterpret_cast<volatile tt_l1_ptr uint32_t*>(page_size_addr)[0] = page_size >> 4;
```

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp (hunk 15 above)
Comment (aliuTT) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:58]: Yea, I had the same thought. technically you can have read ptr not be 16B aligned but the *16 then divide by 16 negates that.

File: tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp (hunk 4 above)
Comment (abhullar-tt) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: can this api be hidden from user and be called when the umd for a mmio device + devices its connected to via ethernet is created or would setting this to `true` break SD flow?

Comment (abhullar-tt) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: ahh yeah sorry rmr this discussion, do you think it makes sense for remote device initialization to call this?

Comment (aliuTT) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: yea probably. Wherever we will use this CreateDevice pattern that is used in gtests.

Comment (abhullar-tt) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: nvm my original comment, if each remote device calls this api it may mess up the state for other already initialized devices

Comment (aliuTT) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: @TT-billteng we'll need (probably) a metal host api that does this:
```
void initialize_my_cluster(vec<Devices> devs) {
    find all required mmio chips and remote chips to support all devs
    for ( all required chips) {
        CreateDevice(required chip);
    }
    tt::Cluster::instance().set_internal_routing_info_for_ethernet_cores(true);
}
```

The main reason is that users may only want to use one remote chip, but at the metal level we need to create and initialize some other devices to support a FD tunneling path to that remote chip.

File: tt_metal/hw/inc/dev_msgs.h (hunk 12 above)
Comment (aliuTT) [reply tt_metal/hw/inc/dev_msgs.h:129]: Implying a field user should not use, sort of like private members I guess

File: tt_metal/hw/firmware/src/erisc.cc [hunk 20]
```diff
@@ -28,13 +28,50 @@ uint32_t wIndex __attribute__((used));
...
+    eth_db_cb_config->ack += num_pages;
+
+    uint32_t pages_ack_addr = (uint32_t)(&(remote_db_cb_config->ack));
+    noc_semaphore_set_remote((uint32_t)(&(eth_db_cb_config->ack)), producer_noc_encoding | pages_ack_addr);
+}
+
+FORCE_INLINE
+uint64_t get_dispatch_core_noc_encoding() {
+    uint32_t dispatch_noc_x = (uint32_t)((routing_info->dispatch_core_xy >> 16) & 0xFFFF);
+    uint32_t dispatch_noc_y = (uint32_t)(routing_info->dispatch_core_xy & 0xFFFF);
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:65]: I left one 16 byte chunk for metadata: 
```
    volatile uint32_t routing_enabled;
    volatile uint32_t routing_mode;
    volatile uint32_t connected_chip_id;
    volatile uint32_t dispatch_core_xy;
    ```
  And I didn't want to use another 16 bytes lol. I'll add more space to this, because I have a feeling we'll need more info here later on anyway

File: tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp (hunk 4 above)
Comment (abhullar-tt) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: yeah I think this is a good idea and it can accept target device ids

Comment (TT-billteng) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: are you saying this is a top-level API in `host_api.hpp`?

sounds like if had a device pool that's always initialized under the hood, this wouldn't need to be another top-level API

Comment (aliuTT) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: I think that is the same discussion as "why do we have users call CreateDevice for single chip today". We are proposing some api that is hot-swappable with the CreateDevice call for single chip.

Comment (TT-billteng) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: if the user needs to only use one chip, I feel like it would be confusing for them to have to figure out which other chips they need to start up to get the remote chip working. This should be infra hidden under the top-level API

Comment (TT-billteng) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: I can imagine we'd want an API that creates a handle to more than 1 or all the devices like `CreateDevices( vector<Device*> dev)` but if user wants access to 1 remote chip only, we should figure out the associated MMIO chip to start up as well

Comment (aliuTT) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: Yea, Almeet and I agree with that. We have to find the user all the chips to initialize, because the FD routing is something tt_metal manages under the hood.

File: tt_metal/hw/inc/ethernet/dataflow_api.h [hunk 21]
```diff
@@ -154,8 +152,12 @@ void eth_send_bytes(
...
 FORCE_INLINE
 void eth_wait_for_receiver_done() {
-    eth_send_packet(0, ((uint32_t)(&(erisc_info->bytes_sent))) >> 4, ((uint32_t)(&(erisc_info->bytes_sent))) >> 4, 1);
-    while (erisc_info->bytes_sent != 0) {
+    eth_send_packet(
+        0,
+        ((uint32_t)(&(erisc_info->user_buffer_bytes_sent))) >> 4,
+        ((uint32_t)(&(erisc_info->user_buffer_bytes_sent))) >> 4,
+        1);
+    while (erisc_info->user_buffer_bytes_sent != 0) {
```
Comment (pgkeller) [inline tt_metal/hw/inc/ethernet/dataflow_api.h:215]: try to split cleanup like this into separate commits from substantive changes, otherwise leaves the reviewers eyes trying to figure out what changed

Comment (aliuTT) [reply tt_metal/hw/inc/ethernet/dataflow_api.h:215]: I agree, I usually try to keep this sort of stuff more organized in my commits, but this PR got too big so I squashed a lot of things 🤡

File: tt_metal/hw/inc/dev_msgs.h [hunk 22]
```diff
@@ -112,9 +112,19 @@ static_assert(MEM_MAILBOX_BASE + offsetof(mailboxes_t, ncrisc_halt.stack_save) =
...
     volatile uint32_t routing_enabled;
     volatile uint32_t routing_mode;
-    volatile uint32_t unused_arg0;
-    volatile uint32_t unused_arg1;
+    volatile uint32_t connected_chip_id;
+    volatile uint32_t dispatch_core_xy;
+    volatile uint32_t fd_buffer_msgs_sent;
+    uint32_t reserved_0_;
+    uint32_t reserved_1_;
+    uint32_t reserved_2_;
```
Comment (pgkeller) [inline tt_metal/hw/inc/dev_msgs.h:129]: what's the padding?  why 6 uint32?
I think we should pack this tight, so use the smallest applicable datatype.  in refererence to the XY noc address comment above, packing those into a 16 byte value and accessing with 8 bit mask/shfit would be efficient (and save an L1 load).

File: tt_metal/hw/inc/ethernet/dataflow_api.h [hunk 23]
```diff
@@ -63,6 +61,63 @@ void check_and_context_switch() {
...
+        ((uint32_t)(&(routing_info->fd_buffer_msgs_sent))) >> 4,
+        1);
+}
+
+}  // namespace internal_
+
+void run_routing() {
+    // TODO: maybe split into two FWs? or this may be better to sometimes allow each eth core to do both send and
+    // receive of fd packets
+    if (routing_info->routing_mode == EthRouterMode::FD_SRC) {
```
Comment (pgkeller) [inline tt_metal/hw/inc/ethernet/dataflow_api.h:108]: routine_info is in L1 and a volatile, right?  If so, load it once and reference the local variable to save the repeated loads

File: tt_metal/impl/dispatch/kernels/command_queue_producer.hpp [hunk 24]
```diff
@@ -172,6 +216,24 @@ void multicore_cb_push_back(uint64_t consumer_noc_encoding, uint32_t consumer_fi
...
+void multicore_cb_push_back_2(
+    db_cb_config_t* db_cb_config,
+    const db_cb_config_t* remote_db_cb_config,
+    uint64_t consumer_noc_encoding,
+    uint32_t consumer_fifo_16b_limit,
+    uint32_t num_to_write) {
+    db_cb_config->recv += num_to_write;
+    db_cb_config->wr_ptr += db_cb_config->page_size * num_to_write;
+
+    if (db_cb_config->wr_ptr >= consumer_fifo_16b_limit) {
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/kernels/command_queue_producer.hpp:164]: put the computation above into a local, save it then compare to save the extra L1 load

File: tt_metal/impl/device/device.hpp [hunk 25]
```diff
@@ -101,8 +101,8 @@ class Device {
 
     std::vector<CoreCoord> ethernet_cores_from_logical_cores(const std::vector<CoreCoord> &logical_cores) const;
 
-    std::unordered_set<chip_id_t> get_ethernet_connected_chip_ids() const {
-        return tt::Cluster::instance().get_ethernet_connected_chip_ids(this->id_);
+    std::unordered_map<chip_id_t, std::unordered_set<CoreCoord>> get_ethernet_cores_grouped_by_connected_chips() const {
```
Comment (pgkeller) [inline tt_metal/impl/device/device.hpp:104]: use a vector

File: tt_metal/llrt/tt_cluster.cpp [hunk 26]
```diff
@@ -607,6 +634,100 @@ std::tuple<chip_id_t, CoreCoord> Cluster::get_connected_ethernet_core(std::tuple
         std::get<0>(connected_eth_core), soc_desc.chan_to_logical_eth_core_map.at(std::get<1>(connected_eth_core)));
 }
 
+CoreCoord Cluster::ethernet_core_from_logical_core(chip_id_t chip_id, const CoreCoord &logical_core) const {
+    const metal_SocDescriptor &soc_desc = get_soc_desc(chip_id);
+    return soc_desc.get_physical_ethernet_core_from_logical(logical_core);
+}
+
+void Cluster::set_routing_config_for_ethernet_cores() {
+    const char *TT_METAL_SLOW_DISPATCH_MODE = std::getenv("TT_METAL_SLOW_DISPATCH_MODE");
```
Comment (pgkeller) [inline tt_metal/llrt/tt_cluster.cpp:627]: beyond the scope of this change, but we should move the env var into the settings and pull that rather than looking up a string...

File: tt_metal/hw/inc/dev_msgs.h (hunk 22 above)
Comment (aliuTT) [reply tt_metal/hw/inc/dev_msgs.h:129]: `fd_buffer_msgs_sent` is getting sent with ethernet hardware which acts on 16 byte chunks. The receiver eth core writes to this address on the sender eth core. I agree about the 8 bit mask/shift for perf optimization

File: tt_metal/hw/inc/ethernet/dataflow_api.h (hunk 23 above)
Comment (aliuTT) [reply tt_metal/hw/inc/ethernet/dataflow_api.h:108]: I'll do that, but better yet, I intended to build an `erisc_app_src.hex` and `erisc_app_dst.hex` at compile time and load different FWs to these eth cores. I believe the routing mode of an eth core should never change. This routing mode toggle is here for bring up and for debugging purposes.

File: tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp (hunk 4 above)
Comment (aliuTT) [reply tests/tt_metal/tt_metal/unit_tests/common/device_fixture.hpp:35]: tracked here: https://github.com/tenstorrent-metal/tt-metal/issues/5048

File: tt_metal/hw/firmware/src/erisc.cc [hunk 27]
```diff
@@ -28,13 +28,50 @@ uint32_t wIndex __attribute__((used));
...
+    return uint64_t(NOC_XY_ENCODING(dispatch_noc_x, dispatch_noc_y)) << 32;
+}
+
+FORCE_INLINE
+void eth_db_acquire(volatile uint32_t *semaphore, uint64_t noc_encoding) {
+    while (semaphore[0] == 0 and routing_info->routing_enabled) {
+    }
```
[the remark this answers is not in the document]
Comment (aliuTT) [reply tt_metal/hw/firmware/src/erisc.cc:72]: Routing enabled is required for SD and FD. We can split the loaded erisc FW for SD, but I'm saving that for the next round of FW changes

File: tt_metal/impl/dispatch/kernels/command_queue_producer.hpp (hunk 24 above)
Comment (aliuTT) [reply tt_metal/impl/dispatch/kernels/command_queue_producer.hpp:164]: `db_cb_config` is a local copy already

File: tt_metal/impl/dispatch/kernels/command_queue_common.hpp [hunk 28]
```diff
@@ -5,56 +5,25 @@
...
+    volatile uint32_t wr_ptr ATTR_ALIGNL1;      // 16B
+};
+static constexpr uint32_t l1_db_cb_addr_offset = sizeof(db_cb_config_t);
 
 template <uint32_t cmd_base_address, uint32_t consumer_data_buffer_size>
 FORCE_INLINE uint32_t get_command_slot_addr(bool db_buf_switch) {
     static constexpr uint32_t command0_start = cmd_base_address;
     static constexpr uint32_t command1_start = command0_start + DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND + consumer_data_buffer_size;
-    return (db_buf_switch) ? command0_start : command1_start;
+    return (not db_buf_switch) ? command0_start : command1_start;
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_common.hpp:26]: good catch

File: tt_metal/impl/dispatch/kernels/command_queue_common.hpp [hunk 29]
```diff
@@ -64,10 +33,23 @@ FORCE_INLINE uint32_t get_db_buf_addr(bool db_buf_switch) {
...
-
 FORCE_INLINE
 void db_acquire(volatile uint32_t* semaphore, uint64_t noc_encoding) {
     while (semaphore[0] == 0);
     noc_semaphore_inc(noc_encoding | uint32_t(semaphore), -1); // Two's complement addition
     noc_async_write_barrier();
 }
+
+// Local refers to the core that is calling this function
+db_cb_config_t* get_local_db_cb_config(uint32_t base_addr, bool db_buf_switch) {
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_common.hpp:44]: you should label this as an L1 pointer (tt_l1_ptr)

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp [hunk 30]
```diff
@@ -35,45 +35,36 @@ void noc_async_write_multicast_one_packet_no_path_reserve(
...
+    db_cb_config_t* db_cb_config,
+    const db_cb_config_t* remote_db_cb_config,
     uint64_t producer_noc_encoding,
-    bool db_buf_switch,
-    uint32_t fifo_limit,
-    uint32_t fifo_size,
-    uint32_t num_pages,
+    uint32_t consumer_fifo_limit,
+    uint32_t num_to_write,
     uint32_t page_size) {
```
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:54]: From the looks of it, consumer_fifo_limit is in terms of bytes, but page_size is 16B. This is confusing, and we need to ensure that we're consistent. Not sure how many cases in the code don't have this inconsistency, but these should be resolved before merging.

**Review by DrJessop — CHANGES_REQUESTED** [review]
I would say the following, to minimize code changes, CB data should be assumed to be 16B aligned for sizes/addresses, but that should not be the case for the APIs.

We can revisit later to explicitly mark 16B data as 16B with a suffix, but I don't think we should change the logic this much right now.

**Review by DrJessop — APPROVED** [review]

**Review by TT-billteng — APPROVED** [review]

**Review by abhullar-tt — APPROVED** [review]
